### PR TITLE
[JENKINS-51170] Enable the usage of the StepEnvironmentContributor extensionpoint

### PR DESCRIPTION
**Pull request by t-8ch**
Support the new StepEnvironmentContributor hook introduced in
jenkinsci/workflow-step-api-plugin#36.

This depends on jenkinsci/workflow-step-api-plugin#36.
The dependency version number used depends on the incrementalifyiation (JEP 305).
If requested I can perform this and submit another PR.

### CODE REVIEW

**Comment by jglick** [comment]
> The dependency version number used depends on the incrementalifyiation (JEP 305).
> If requested I can perform this

Yes should suffice to update the parent (currently to 3.12) and

```bash
mvn incrementals:incrementalify
```

Could be done in this PR so far as I am concerned.

**Comment by jglick** [comment]
(Somehow the CI build seems to be running even without the `consume-incrementals` profile, not sure why. Might be an artifact of mirror setup on the CI builder.)

**Comment by t-8ch** [comment]
I got an error an error with access restrictions (DoNotUse).
Will try again next week, just left for the weekend.

**Comment by t-8ch** [comment]
The incrementalization is blocked on https://github.com/kohsuke/access-modifier/pull/12
(Or we work around by removing some access modifiers)

**Comment by t-8ch** [comment]
The hooks work so far for steps. The logic in https://github.com/jenkinsci/workflow-support-plugin/pull/62 is actually sufficient. This PR only contains tests.

I am not sure about also calling the hooks for `EnvActionImpl`.
This would require to get the current interpreter state from groovy, getting the enclosing scope and somehow using this to fetch the parents scope `StepContext` from somewhere inside the CPS engine.

Any opinions if this part is required/wanted/superfluous/etc?

File: pom.xml [hunk 1]
```diff
@@ -28,12 +28,12 @@
     <parent>
         <groupId>org.jenkins-ci.plugins</groupId>
         <artifactId>plugin</artifactId>
-        <version>3.4</version>
+        <version>3.12</version>
```
Comment (jglick) [inline pom.xml]: Recommend merging with #230 to avoid conflicts.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/EnvActionImpl.java [hunk 2]
```diff
@@ -88,7 +96,12 @@ private EnvActionImpl() {
     @Override public String getProperty(String propertyName) {
         try {
             CpsThread t = CpsThread.current();
-            return EnvironmentExpander.getEffectiveEnvironment(getEnvironment(), t.getContextVariable(EnvVars.class), t.getContextVariable(EnvironmentExpander.class)).get(propertyName);
+            TaskListener listener = getListener();
+
+            return EnvironmentExpander.getEffectiveEnvironment(
+                    getEnvironment(getListener()), t.getContextVariable(EnvVars.class),
+                    t.getContextVariable(EnvironmentExpander.class), null, listener)
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/EnvActionImpl.java]: Note that if you follow my suggestion in the API PR, this call will then have identical behavior to what it did before—just avoiding a deprecation warning.

(I see no reason for `env.SOMETHING` to behave differently. There is no particular `StepContext` when you are in the middle of Groovy code. There are contexts for enclosing blocks, but the API is defined to allow setting an environment for a step, not for a chain of blocks.)

File: src/test/java/org/jenkinsci/plugins/workflow/DynamicEnvironmentExpanderTest.java [hunk 3]
```diff
@@ -145,4 +152,64 @@
         }
     }
 
+    @Issue("JENKINS-51170")
+    @Test public void perStepEnvironment() {
+        story.addStep(new Statement() {
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/DynamicEnvironmentExpanderTest.java]: BTW we are on Java 8 here now, so you can use

```java
story.then(r -> {
    WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "p");
    // …
});
```

File: src/test/java/org/jenkinsci/plugins/workflow/DynamicEnvironmentExpanderTest.java [hunk 4]
```diff
@@ -145,4 +152,64 @@
...
+            @Override public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+                // p.setDefinition(new CpsFlowDefinition("echo(/before VAR=$VAR/); " + StepEnvAdder.class.getCanonicalName() + ".value = 'after'; echo(/after VAR=$VAR/)", false));
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/DynamicEnvironmentExpanderTest.java]: Any reason why the commented-out code is still here? Left over from some debugging session?

File: src/test/java/org/jenkinsci/plugins/workflow/DynamicEnvironmentExpanderTest.java [hunk 5]
```diff
@@ -145,4 +152,64 @@
...
+                StepContext context = getContext();
+                String message = this.var + "=" + context.get(EnvVars.class).get(var);
+                context.get(TaskListener.class).getLogger().println(message);
+                return null;
+            }
+        }
+
+        @TestExtension("perStepEnvironment") public static class DescriptorImpl extends StepDescriptor {
+            @Override public String getFunctionName() { return "printEnv"; }
+            @Override public Set<? extends Class<?>> getRequiredContext() { return Collections.emptySet(); }
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/DynamicEnvironmentExpanderTest.java]: actually

```java
return ImmutableSet.of(EnvVars.class, TaskListener.class);
```

File: src/main/java/org/jenkinsci/plugins/workflow/cps/EnvActionImpl.java (hunk 2 above)
Comment (t-8ch) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/EnvActionImpl.java]: @jglick Ok. Personally I only need the environment inside the steps. I assumed you wanted to keep the behaviour consistent. Let's leave it like that then.

**Comment by t-8ch** [comment]
@jglick I implemented your comments. Thanks for the review!

**Comment by t-8ch** [comment]
I see the following alternatives to aquire a listener:

1) `getExecution().getOwner().getListener()`
  * Is used just above
2) Remove the check that is throwing the exception (only for our internal call)
  * May have side-effects
3) Do the same as `EnvActionImpl.getListener()`
  * Looks to be similar to 1), but falls back to plain logging

Personally I think 3) seems to be the best.

**Comment by t-8ch** [comment]
I implemented prototypes for all three scenarios and all of them pass the testsuites of `workflow-support` and `workflow-cps`.

**Comment by dwnusbaum** [comment]
Thanks for looking into all of the options! Given that the line immediately above the new line in `workflow-support` uses `getExecution().getOwner().getListener()`, option 1 seems the best to me for consistency with that behavior. Option 2 seems a little worrying because of potential side effects like you noted, and it seems like if we used option 3, anything sent to the system logs would probably just end up being unhelpful log spam since it wouldn't be associated with the particular build it came from. So my preference would be 1 over 3 over 2.

**Comment by t-8ch** [comment]
Will do. I won't get around until tomorrow though. If you want feel free to take a shot at it.

**Comment by jglick** [comment]
> getExecution().getOwner().getListener()

Sounds good but better do it in `DefaultStepContext.getListener`: https://github.com/jenkinsci/workflow-support-plugin/pull/62#discussion_r248848737

> Do the same as `EnvActionImpl.getListener()`

No, do not do that.

**Comment by t-8ch** [comment]
I went with 2 with a fallback to 1, as recommended by @jglick . The tests seem to be happy.

**Pull request merged**